File: r/tests/testthat/test_voter_sampler.R
# test_voter_sampler.R — Tests for make_voter_sampler() and draw_voters().

test_that("make_voter_sampler returns correct structure for gaussian", {
  s <- make_voter_sampler("gaussian", mean = 0.5, sd = 0.3)
  expect_s3_class(s, "VoterSamplerConfig")
  expect_equal(s$kind,   1L)
  expect_equal(s$param1, 0.5)
  expect_equal(s$param2, 0.3)
})

test_that("make_voter_sampler returns correct structure for uniform", {
  s <- make_voter_sampler("uniform", lo = -2.0, hi = 2.0)
  expect_s3_class(s, "VoterSamplerConfig")
  expect_equal(s$kind,   0L)
  expect_equal(s$param1, -2.0)
  expect_equal(s$param2,  2.0)
})

test_that("make_voter_sampler errors on bad sd", {
  expect_error(make_voter_sampler("gaussian", sd = 0))
  expect_error(make_voter_sampler("gaussian", sd = -1))
})

test_that("make_voter_sampler errors when lo >= hi", {
  expect_error(make_voter_sampler("uniform", lo = 1, hi = 1))
  expect_error(make_voter_sampler("uniform", lo = 2, hi = 1))
})

test_that("draw_voters returns numeric vector of correct length (gaussian)", {
  skip_without_lib()
  sm <- stream_manager(42L, "vs")
  s  <- make_voter_sampler("gaussian", mean = 0, sd = 0.4)
  v  <- draw_voters(50L, s, sm, "vs")
  expect_type(v, "double")
  expect_length(v, 100L)
  expect_true(all(is.finite(v)))
})

test_that("draw_voters returns numeric vector of correct length (uniform)", {
  skip_without_lib()
  sm <- stream_manager(7L, "vs")
  s  <- make_voter_sampler("uniform", lo = -1, hi = 1)
  v  <- draw_voters(30L, s, sm, "vs")
  expect_type(v, "double")
  expect_length(v, 60L)
  expect_true(all(v >= -1 & v <= 1))
})

test_that("draw_voters is reproducible given same seed", {
  skip_without_lib()
  s    <- make_voter_sampler("gaussian", mean = 0, sd = 1)
  sm1  <- stream_manager(99L, "vs")
  sm2  <- stream_manager(99L, "vs")
  expect_equal(draw_voters(20L, s, sm1, "vs"),
               draw_voters(20L, s, sm2, "vs"))
})

test_that("draw_voters result is usable as voter_ideals in competition_run", {
  skip_without_lib()
  sm      <- stream_manager(1L, c("vs", "competition_adaptation_hunter"))
  s       <- make_voter_sampler("gaussian", mean = 0, sd = 0.4)
  voters  <- draw_voters(30L, s, sm, "vs")
  headings <- c(0.71, 0.71, -0.71, -0.71)
  trace <- competition_run(
    competitor_positions = c(-0.5, -0.5, 0.5, 0.5),
    competitor_headings  = headings,
    strategy_kinds       = c("hunter", "hunter"),
    voter_ideals         = voters,
    dist_config          = make_dist_config(n_dims = 2L),
    engine_config        = make_competition_engine_config(
      seat_count = 1L,
      seat_rule  = "plurality_top_k",
      max_rounds = 5L,
      step_config = make_competition_step_config(
        kind = "fixed", fixed_step_size = 0.1
      )
    ),
    stream_manager = sm
  )
  expect_s3_class(trace, "CompetitionTrace")
  expect_gte(trace$dims()$n_rounds, 1L)
})

test_that("draw_voters errors when sampler is not VoterSamplerConfig", {
  skip_without_lib()
  sm <- stream_manager(1L, "vs")
  expect_error(draw_voters(10L, list(kind = 0L), sm, "vs"))
})

File: python/tests/test_voter_sampler.py
import numpy as np
import pytest

import socialchoicelab as scl
from socialchoicelab import VoterSamplerConfig, draw_voters, make_voter_sampler


# ==========================================================================
# make_voter_sampler
# ==========================================================================


def test_make_voter_sampler_gaussian_config():
    s = make_voter_sampler("gaussian", mean=0.5, stddev=0.3)
    assert isinstance(s, VoterSamplerConfig)
    assert s.kind == 1
    assert s.param1 == pytest.approx(0.5)
    assert s.param2 == pytest.approx(0.3)


def test_make_voter_sampler_uniform_config():
    s = make_voter_sampler("uniform", lo=-2.0, hi=2.0)
    assert isinstance(s, VoterSamplerConfig)
    assert s.kind == 0
    assert s.param1 == pytest.approx(-2.0)
    assert s.param2 == pytest.approx(2.0)


def test_make_voter_sampler_unknown_kind_raises():
    with pytest.raises(ValueError, match="unknown kind"):
        make_voter_sampler("dirichlet")


# ==========================================================================
# draw_voters
# ==========================================================================


def test_draw_voters_gaussian_length():
    sm = scl.StreamManager(master_seed=42, stream_names=["vs"])
    s = make_voter_sampler("gaussian", mean=0.0, stddev=0.4)
    v = draw_voters(50, s, sm, "vs")
    assert isinstance(v, np.ndarray)
    assert v.dtype == np.float64
    assert len(v) == 100
    assert np.all(np.isfinite(v))


def test_draw_voters_uniform_in_range():
    sm = scl.StreamManager(master_seed=7, stream_names=["vs"])
    s = make_voter_sampler("uniform", lo=-1.0, hi=1.0)
    v = draw_voters(200, s, sm, "vs")
    assert len(v) == 400
    assert np.all(v >= -1.0)
    assert np.all(v <= 1.0)


def test_draw_voters_reproducible():
    s = make_voter_sampler("gaussian", mean=0.0, stddev=1.0)
    sm1 = scl.StreamManager(master_seed=99, stream_names=["vs"])
    sm2 = scl.StreamManager(master_seed=99, stream_names=["vs"])
    np.testing.assert_array_equal(
        draw_voters(30, s, sm1, "vs"),
        draw_voters(30, s, sm2, "vs"),
    )


def test_draw_voters_usable_as_voter_ideals():
    """Voters from draw_voters should feed directly into competition_run."""
    sm = scl.StreamManager(
        master_seed=1, stream_names=["vs", "competition_adaptation_hunter"]
    )
    s = make_voter_sampler("gaussian", mean=0.0, stddev=0.4)
    voters = draw_voters(30, s, sm, "vs")
    # Reshape to (n_voters, n_dims) for competition_run
    voter_matrix = voters.reshape(-1, 2)

    competitors = np.array([[-0.5, -0.5], [0.5, 0.5]])
    headings = np.array([0.71, 0.71, -0.71, -0.71])
    cfg = scl.CompetitionEngineConfig(
        seat_count=1,
        seat_rule="plurality_top_k",
        max_rounds=5,
        step_config=scl.CompetitionStepConfig(kind="fixed", fixed_step_size=0.1),
    )
    with scl.competition_run(
        competitors,
        ["hunter", "hunter"],
        voter_matrix,
        dist_config=scl.DistanceConfig(salience_weights=[1.0, 1.0]),
        engine_config=cfg,
        stream_manager=sm,
    ) as trace:
        n_rounds, n_competitors, n_dims = trace.dims()
        assert n_competitors == 2
        assert n_rounds >= 1
        assert n_dims == 2


def test_draw_voters_wrong_sampler_type_raises():
    sm = scl.StreamManager(master_seed=1, stream_names=["vs"])
    with pytest.raises(TypeError, match="VoterSamplerConfig"):
        draw_voters(10, {"kind": 0, "param1": -1.0, "param2": 1.0}, sm, "vs")
